### PR TITLE
lstat before readlink (#44)

### PR DESCRIPTION
**Pull request by aremmell**

### CODE REVIEW

**Review by johnsonjh — APPROVED** [review]
Looks fine - I didn't get a chance to test non-Linux however

**Comment by aremmell** [comment]
Btw, I see why that `grow = false` got marked as 'dead code': it's because the do/while exits if grow is set to anything other than true, so basically is an unnecessary variable

File: sirfilesystem.c [hunk 1]
```diff
@@ -158,42 +158,43 @@ char* _sir_getcwd(void) {
...
+# elif defined(__SOLARIS__)
+#  define PROC_SELF "/proc/self/path/a.out"
+# endif
+    struct stat st;
+    if (-1 == lstat(PROC_SELF, &st)) {
         _sir_handleerr(errno);
         return NULL;
     }
 
+    size_t size = (st.st_size > 0) ? st.st_size + 1 : SIR_MAXPATH;
```
Comment (johnsonjh) [inline sirfilesystem.c:179]: Looks like you need an exception for 0 too:

Like the example code in https://github.com/aremmell/libsir/issues/44#issuecomment-1625511800 -
```c
           /* Some magic symlinks under (for example) /proc and /sys
              report 'st_size' as zero. In that case, take PATH_MAX as
              a "good enough" estimate. */

           if (sb.st_size == 0)
               bufsiz = PATH_MAX;
```

**Review by johnsonjh — APPROVED** [review]
Actually looks 👍🏻 

I'll run through the other OSes if you want tho but I think it's fine

**Comment by aremmell** [comment]
> Actually looks 👍🏻
> 
> I'll run through the other OSes if you want tho but I think it's fine

Might as well try one, just to make sure I didn't explode it. Travis is checking Linux, FreeBSD and macOS

**Comment by aremmell** [comment]
I'm going to check Windows again. It usually gets broken every 5-10 changes, and Travis' windows support is laughable.

**Comment by johnsonjh** [comment]
@aremmell well, some issues 

Solaris and illumos is fine, DragonFly failed:

IMG_7864

didn't get to try NetBSD or Cygwin yet

**Review by johnsonjh — CHANGES_REQUESTED** [review]
DragonFly failed

**Comment by aremmell** [comment]
And we're 100% sure that worked on DragonFly before? Because to my eyes, the code didn't really change, it just doesn't try to re-grow the buffer. Let me throw some _sir_selflogs in there.

**Comment by johnsonjh** [comment]
> And we're 100% sure that worked on DragonFly before? Because to my eyes, the code didn't really change, it just doesn't try to re-grow the buffer. Let me throw some _sir_selflogs in there.

Yep, with master:
> done: all 25 libsir tests passed in 1.416sec!

**Comment by johnsonjh** [comment]
Same failure on NetBSD ...

**Comment by aremmell** [comment]
Okay, well, can you confirm in the man pages for readlink that if it returns the same value as the size value you sent in, that it could've overflowed?

Because if it does, it's lying (or lstat is), and we'll have to treat it specially?

Anyway, pull again and let's see what selflog has to tell us.

**Comment by johnsonjh** [comment]
Both claim POSIX conformance 

>   The readlink() function conforms to IEEE Std 1003.1-2001 ("POSIX.1").

**Comment by johnsonjh** [comment]
IMG_7865

Cwd is "/home/jhj/src/libsir" so 

"/home/jhj/src/libsir/build/bin/sirtests" is 39 + 1 = 40 (41 if null terminated)

**Comment by aremmell** [comment]
Ok, I know what's going on. I'm passing size - 1 which defeats the purpose of adding 1 to begin with. Have to add 2, because we need that null term.

**Comment by johnsonjh** [comment]
Linux is just falling back


_sir_getappfilename (sirfilesystem.c:180): lstat(/proc/self/exe) = st.st_size = 0
_sir_getappfilename (sirfilesystem.c:189): allocating 4096 bytes for filename
_sir_getappfilename (sirfilesystem.c:201): readlink(/proc/self/exe) returned 41, passed in size 4095
        _sir_pathexists('file.exists') = true

seems like this might be kind of a waste on Linux, ironically!  But helpful on Solaris and the BSDs.

**Comment by johnsonjh** [comment]
Also, above the actual readlink call, need to add `/* Flawfinder: ignore */` - it's not smart enough to know the mitigation is in place.

**Comment by aremmell** [comment]
Sometimes I really hate git. I just accidentally committed changes not relevant to this; Windows (and clang-analyzer) gave me shit about a couple things so I squashed them.

Go ahead and pull and try again.

**Comment by johnsonjh** [comment]
@aremmell Very good; NetBSD and DragonFly now passing.

**Review by johnsonjh — APPROVED** [review]
Tested and working on Cygwin, Linux, Solaris, NetBSD, DragonFly, illumos.

**Comment by johnsonjh** [comment]
FreeBSD and Mac don't use readlink here so it should be fine, but Travis will test those also

**Comment by aremmell** [comment]
> FreeBSD and Mac don't use readlink here so it should be fine, but Travis will test those also

I don't think Travis is doing a damn thing right now.

**Comment by aremmell** [comment]
There we go. I merged your ignore *.d PR and it was building that still.

**Comment by johnsonjh** [comment]
I see FreeBSD is OK; waiting on mac

**Pull request merged**